File: 백준/Gold/1245. 농장 관리/농장 관리.py
def is_top(r, c):  # 현재위치의 높이의 집합이 봉우리인지 확인
    now_height = farm[r][c]
    stack = [(r, c)]
    flag = True
    while stack:
        r, c = stack.pop()
        for dr, dc in drc:
            nr = r + dr
            nc = c + dc
            if 0 <= nr < N and 0 <= nc < M:
                if not visited[nr][nc] and farm[nr][nc] == now_height:
                    stack.append((nr, nc))  # 1. 인접한 동일 높이는 모두 탐색함
                    visited[nr][nc] = 1  # 2. 방문체크
                elif farm[nr][nc] > now_height:  # 1. 인접한 격자 중에 현재보다 높은 곳이 존재할 경우 봉우리가 아니다
                    flag = False  # 봉우리 체크

    if flag:
        return 1

    return 0


N, M = map(int, input().split())
farm = [list(map(int, input().split())) for _ in range(N)]

visited = [[0] * M for _ in range(N)]

drc = [[0, -1], [-1, -1], [-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1]]
cnt = 0
stack = []

for r in range(N):
    for c in range(M):
        if not visited[r][c] and is_top(r, c):
            cnt += 1

print(cnt)

File: 백준/Gold/1245. 농장 관리/농장 관리.cc
#include <iostream>
#define MAX_VAL 101
#include <stack>

using namespace std;

int n, m;
int farm[MAX_VAL][MAX_VAL];
int visited[MAX_VAL][MAX_VAL];
int dx[] = {0, -1, -1, -1, 0, 1, 1, 1};
int dy[] = {-1, -1, 0, 1, 1, 1, 0, -1};

bool is_top(int r, int c)
{
  int now_height = farm[r][c];
  stack<pair<int, int>> stack;
  bool flag = true;

  pair<int, int> tmp;

  stack.push(make_pair(r, c));

  while (!stack.empty()) {
    tmp.first = stack.top().first;
    tmp.second = stack.top().second;
    stack.pop();

    for (int i=0; i<8; i++){
      int nr = tmp.first + dx[i];
      int nc = tmp.second + dy[i];
      if (0 <= nr && nr < n && 0 <= nc && nc < m){
        if (!visited[nr][nc] && farm[nr][nc] == now_height) {
          stack.push(make_pair(nr, nc));
          visited[nr][nc] = 1;
        }
        else if (farm[nr][nc] > now_height)
        {
          flag = false;
        }
        
      }
    }
  }
  return flag;
}

int main()
{
  ios_base::sync_with_stdio(false);cin.tie(nullptr);cout.tie(nullptr);

  cin >> n >> m;

  for (int i=0; i<n; i++) {
    for(int j=0; j<m; j++) {
      cin >> farm[i][j];
      visited[i][j] = 0;
    }
  }
  
  int cnt = 0;

  for (int i=0; i<n; i++){
    for (int j=0; j<m; j++){
      if (!visited[i][j] && is_top(i, j)){
        cnt ++;
      }
    }
  }

  cout << cnt << endl;

  return 0;
}